### ISSUE TITLE
Bug: Metabase truncates CSV/xlsx BigQuery downloads with large data volume.

### ISSUE THREAD

**Issue by hden** [metabase/metabase#6960]
Metabase truncates CSV/xlsx downloads with large data volume.

Minimal required SQL
```sql
#standardsql
select *
from `bigquery-public-data.samples.github_timeline`
limit 20000
```

`wc -l query_result_2018-02-15T09_17_51.084Z.csv` expected 20,000 rows, actual 1,759 rows in our system.

-------
- Your browser and the version: Chrome 63.0
- Your operating system: OS X 10
- Your databases: BigQuery
- Metabase version: v0.28.0
- Metabase hosting environment: jar
- Metabase internal database: MySQL
-------

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by hden** [comment]
@salsakran Any update on this?

**Comment by almegeddon** [comment]
We're seeing the same issue when trying to download large-ish .csv files (~77k rows, well under the 1 million row limit for CSVs).

Digging into the BigQuery driver (bigquery.clj) it looks like the `post-process-native` function is not properly handling paged results. I hacked in a this logging statement to see if I could prove this theory:

```clojure
(log/info (u/format-color 'cyan (str "BigQuery QueryResponse details, Total rows: " (.getTotalRows response) " Fetched Rows: " (.size (.getRows response)) " Page Token: " (.getPageToken response))))
```

and I'm seeing fetched rows less than the total rows available when downloading full results via .csv:

```shell
03-05 13:44:03 INFO driver.bigquery :: BigQuery QueryResponse details, Total rows: 77699 Fetched Rows: 31818 Page Token: BFSFOYSPNEAQAAASA4EAAEEAQCAAKGQIBDFPQAIQUCGQMIFQVYKQ====
```

The `process-native & post-process-native` functions need to check the response for a page token and paginate the results accordingly when present. My guess is in my case (as well as @hden's), the result set is truncated due to BQs Query Job default results size limit [Jobs:query.maxResults](https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/query):

> [Optional] The maximum number of rows of data to return per page of results. Setting this flag to a small value such as 1000 and then paging through results might improve reliability when the query result set is large. In addition to this limit, responses are also limited to 10 MB. By default, there is no maximum row count, and only the byte limit applies.

This issue is a blocker for BigQuery - even when results are limited by the UI's imposed 10k limit. Depending on the content of your results, it is entirely feasible for a result set with less than 10k rows getting truncated.

**Comment by setyven** [comment]
any visibility when this will be looked into / fixed?

**Comment by almegeddon** [comment]
I've implemented changes locally to page results for csv/xlsx/json downloads. Hoping to create a pull request today.

**Comment by setyven** [comment]
Nice! hopefully it can be added [here](https://github.com/metabase/metabase/milestone/72) and make it to 0.32 release

**Comment by jlisyc** [comment]
related https://github.com/metabase/metabase/issues/9651

**Comment by Rohith6215** [comment]
@setyven Is this gonna be resolved on the 0.32 release ?

**Comment by jshafe** [comment]
Would be great to get this fixed as the tool is potentially deceptive and not very useful as is.

**Comment by jshafe** [comment]
Missed the update by 4 hours. Thank you very much!

**Comment by hden** [comment]
AFAIK #6661 is not merged yet. Can anyone confirm the fix? @almegeddon ?

**Comment by that0n3guy** [comment]
I'm on .32.8 (latest at this time) and this is still an issue with BigQuery.   

Example: I have 11444 rows (I can export them from bigquery console and get all of them) and metabase will only export 4430.

**Comment by dannyeuu** [comment]
At version ` v0.32.10` we have the issue that larger data +~100.000 the process broke and we have to restart the service =(

**Comment by jlisyc** [comment]
Could you fix it in 032.11? 
In production 0.33 is a big challenge by now for some users.

Thanks a lot.

**Comment by mazameli** [comment]
@jlisyc do you want to tell me more about challenges you're having with 0.33 in production? maz@metabase.com

**Comment by jlisyc** [comment]
Ok.. "challenge" could not be the word ... We need a more stable version in production... and by now, 0.32.10 (or 0.32.11) should be that one.  Users are used to manage question in the "old mode".

And by now, and in relation with this issue, they need to download reports with around 300000 rows "with no download errors".

**Comment by rodra-go** [comment]
Please fix this, guys!!!

**Comment by salsakran** [comment]
bumping this in priority given the number of people hitting it.

**Comment by Tbodewes** [comment]
We are having a similar issue with a Snowflake DB. In particular, when we export the results of a query with much more than 10K rows (e.g. 100K), the export will have the correct number of rows, but with a lot of NULL values after a certain point. 

In particular, exporting the results of the below query to CSV works fine:
```
SELECT 
   *
FROM table
WHERE var IS NOT NULL
LIMIT 10000
```
while exporting the same query with `LIMIT 50000` will have table.var NULL very often after e.g. the 25,000th row, even though we explicitly excluded rows with NULL values. For multiple runs of the same query, the problem always appears around the same row number, but not at the exact same row. 

The problem goes away again if we use e.g. `LIMIT 30000`, even though the problem originally appeared for rows 25,000-30,000. The issue also seems to be affected by the number of columns. Selecting a single column eliminates the problem.

Running the same query in Snowflake directly and exporting the results does not give any problems.

The reason for us to want to export relatively large files is to put them in an Excel report for a situation where we cannot easily and securely expose a Metabase dashboard. Because I want our field personnel to be able to update the report with the tool that they are already using for BI, I'd prefer having the export be possible through Metabase.

**Comment by flamber** [comment]
@Tbodewes 
That looks like a specific driver issue with Snowflake (bug SNOW-80208), where Metabase 0.33.6 uses driver version 3.8.0, but the fix is included in driver version 3.8.2.
Metabase 0.34.0 will use Snowflake driver 3.10.2 (ref #9917)

**Comment by erwan-boulard** [comment]
> I've implemented changes locally to page results for csv/xlsx/json downloads. Hoping to create a pull request today.

@almegeddon , is it possible to share the fix ? Thks

**Comment by phonezawphyo** [comment]
I'm still having the problem on v 0.35.3

**Comment by angelinemarilao07** [comment]
encountering this problem on Version 0.35.4

**Comment by sbelak** [comment]
@angelinemarilao07, @phonezawphyo which DB are you using?

**Comment by angelinemarilao07** [comment]
> @angelinemarilao07, @phonezawphyo which DB are you using?

Im using BigQuery.

**Comment by danielwiesner1** [comment]
I'm having this same issue with a query that should result in 78,000 rows but only returns 70,000 rows when downloaded either as a CSV or as an Excel file. Here is what I'm using:

Your browser and the version: Chrome Version 84.0.4147.125 (Official Build)
Your operating system: OS X 10.15.5 
Your databases: BigQuery
Metabase version: v0.36.0

I would appreciate any guidance on how to deal with this problem.

**Comment by snarfed** [comment]
we're seeing this too, on Enterprise 1.36.4. CSV downloads of PostgreSQL queries work fine, but some BigQuery downloads get truncated.

interestingly, for a while we saw the exact same result count in these truncated downloads, 20588, as mentioned in https://github.com/metabase/metabase/issues/6661. it's different now after i tweaked the question config a bit, though.

**Comment by camsaul** [comment]
Fixed by #13530
